### PR TITLE
Undo command, redo command, history command and bug fix

### LINKED ISSUE #68: As a HR manager, I can undo my command
...so that I can easily correct mistakes or revert changes.

### PR DESCRIPTION
**Pull request by yiwern5**
- added undo command 
  - bug1: unable to undo assign task and assign another task to the same person, possibly due to task is not written to the storage yet but person is busy (this.task != null)
   - bug2: unable to undo mark task command
- added redo command
- added history command
- tested for undo, redo and history
- fixed delete command
- fixed mark task command

### CODE REVIEW

File: src/main/java/seedu/address/logic/LogicManager.java [hunk 1]
```diff
@@ -31,31 +31,44 @@ public class LogicManager implements Logic {
...
+
+        if (addressBookModified) {
+            logger.info("Address book modified, saving to file.");
+            try {
+                storage.saveAddressBook(model.getAddressBook());
+            } catch (AccessDeniedException e) {
+                throw new CommandException(String.format(FILE_OPS_PERMISSION_ERROR_FORMAT, e.getMessage()), e);
+            } catch (IOException ioe) {
+                throw new CommandException(String.format(FILE_OPS_ERROR_FORMAT, ioe.getMessage()), ioe);
+            }
```
Comment (howenyap) [inline src/main/java/seedu/address/logic/LogicManager.java:71]: Nice work catching different errors!

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Reverts the {@code model}'s address book to its previous state.
+ */
+public class UndoCommand extends Command {
+
+    public static final String COMMAND_WORD = "undo";
+    public static final String MESSAGE_SUCCESS = "Undo success!";
+    public static final String MESSAGE_FAILURE = "No more commands to undo!";
+
```
Comment (howenyap) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java:18]: Good practice grouping text outputs into constants!

**Review by howenyap — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/model/AddressBook.java [hunk 3]
```diff
@@ -112,12 +119,16 @@ public void setPerson(Person target, Person editedPerson) {
      */
     public void removePerson(Person key) {
         persons.remove(key);
-        tasks.remove(key.getTask());
+        Task task = key.getTask();
+        if (task != null) {
```
Comment (ziyi22) [inline src/main/java/seedu/address/model/AddressBook.java:123]: Good to handle null pointer exception when attempting to remove the task associated with the person.

**Pull request merged**